accord-core/src/main/java/accord/local/Commands.java

```diff
@@ -610,247 +665,236 @@ private static boolean maybeExecute(SafeCommandStore safeStore, SafeCommand safe
-            logger.trace("{}: {} executes after us. Stop listening and removing from waiting on apply set.", command.txnId(), dependency.txnId());
-            waitingOn.removeWaitingOn(dependency.txnId(), dependency.executeAt());
-            safeDependency.removeListener(command.asListener());
-            return true;
-        }
-        else if (dependency.hasBeen(Applied))
-        {
-            logger.trace("{}: {} has been applied. Stop listening and removing from waiting on apply set.", command.txnId(), dependency.txnId());
-            waitingOn.removeWaitingOn(dependency.txnId(), dependency.executeAt());
-            safeDependency.removeListener(command.asListener());
-            return true;
-        }
-        else if (command.isWaitingOnDependency())
-        {
-            logger.trace("{}: adding {} to waiting on apply set.", command.txnId(), dependency.txnId());
-            waitingOn.addWaitingOnApply(dependency.txnId(), dependency.executeAt());
-            waitingOn.removeWaitingOnCommit(dependency.txnId());
-            return false;
-        }
-        else
+        return setTruncated(safeStore, safeCommand, Truncate.TRUNCATE, true);
+    }
+
+    public static Command setTruncated(SafeCommandStore safeStore, SafeCommand safeCommand, Truncate truncate, boolean notifyListeners)
+    {
+        Command command = safeCommand.current();
+        Invariants.checkState(!command.is(Truncated));
+
+        //   1) a command has been applied; or
+        //   2) has been coordinated but *will not* be applied (we just haven't witnessed the invalidation yet); or
+        //   3) a command is durably decided and this shard only hosts its home data, so no explicit truncation is necessary to remove it
+        // TODO (desired): consider if there are better invariants we can impose for undecided transactions, to verify they aren't later committed (should be detected already, but more is better)
+        Invariants.checkState(command.hasBeen(Applied) || !command.hasBeen(PreCommitted) || command.partialTxn().keys().isEmpty());
+
+        Command.Truncated result = truncated(command);
+        safeCommand.set(result);
+        safeStore.progressLog().clear(safeCommand.txnId());
+        if (notifyListeners)
+            safeStore.notifyListeners(safeCommand);
+
+        if (truncate == Truncate.ERASE)
+            safeStore.erase(safeCommand);
+        return result;
+    }
+
+    public static Truncate shouldTruncate(SafeCommandStore safeStore, Command command)
+    {
+        if (safeStore.commandStore().globalDurability(command.txnId()) == Universal)
+            return Truncate.ERASE;
+
+        if (!command.hasBeen(Applied) || !command.known().hasCompleteRoute())
+            return Truncate.NO;
+
+        TxnId txnId = command.txnId();
+        Timestamp executeAt = command.executeAt() != null && !command.executeAt().equals(Timestamp.NONE) ? command.executeAt() : txnId;
```

Comment:
this is a super common pattern... "should" we push this into `Command`? 

```
public Timestamp executeAtOrTxnId()
{
    return executeAt() != null && !executeAt().equals(Timestamp.NONE) ? executeAt() : txnId;
}
```

Might be not grasping the subtle differences, but we have other patterns like `Timestamp executeAt = full.executeAt != null ? full.executeAt : command.known().executeAt.hasDecidedExecuteAt() ? command.executeAt() : txnId;`... so reusing might be good

Reply:
They're two different things you reference, but yes, we can name them in Command to make the distinction and call-sites clearer.